Shift3/boilerplate-client-react
# Adds in-app notifications

## Changes
1. Adds a model for notifications
2. Adds a notificationApi file for the associated requests
3. Adds an `isnull` filter op
4. Adds a base Notification component and one notification component that makes use of it
5. Adds a Notifications button with the unread count in the side bar
6. Adds a `NotificationPage`. It includes sub-tabs that will list the unread and read notifications, respectively. You can individually mark notifications as read, and you can click on the "Mark All Read" button in order to mark them all as read.
7. Adds a `NotificationScrollView` component. This is how the notifications are shown on the NotificationPage. This component uses an infinite scrolling mechanism to load the data. The newly added `useInfiniteLoading` hook supports this operation.
8. Uses Server-Side Events via the EventSource API in order to recognize when new notifications come in. When this happens, the notification indicator in the side bar will be updated accordingly.

## Purpose
Adds in-app notifications

## Approach
I originally planned on using class components for the notification components. However, our use of hooks via redux-toolkit stopped me from going in that direction once I learned that you can't use hooks in class components.

## Learning
- Learned how to implement infinite scrolling. I used the following resources: [How to Implement Infinite Scrolling with React.js](https://www.pluralsight.com/guides/how-to-implement-infinite-scrolling-with-reactjs), [React Hooks for infinite scroll: An advanced tutorial](https://blog.logrocket.com/react-hooks-infinite-scroll-advanced-tutorial/). 
- It is no longer being used, but I learned how to do use redux-toolkit `pollingInterval` property. The following webpage was very useful: [Polling | Redux Toolkit](https://redux-toolkit.js.org/rtk-query/usage/polling)

## Testing Steps
1. Pull in the changes to your local copy of this branch and run it alongside the `feature/notifications` branch of the dj_starter_demo repo.
2. Create 11 agents in order to be able to test the infinite scrolling functionality. The backend will generate agent creation notifications. The unread count will appear when there are at least one unread notification.
3. Click on the Notification button in order to go to the Notification page. Mark a notification as read. Additionally, click on the Mark All Read button when it isn't disabled.

## Screenshots

![notifications - update - unread](https://user-images.githubusercontent.com/2876874/181386804-86f6b977-2be7-4a35-bcd1-32142ad2f01e.png)

![notifications - update - read](https://user-images.githubusercontent.com/2876874/181386819-5a8af139-6796-40f2-98c9-0d904a9fc93b.png)

![Agent Created Notification](https://user-images.githubusercontent.com/2876874/184458052-d4a99372-31d5-4d83-90c0-27acffc91ef9.png)
